File: Ruby/voice/export_calls_to_csv.rb
require "signalwire/sdk"
require "csv"
require "dotenv"

Dotenv.load("./../.env")

@client = Signalwire::REST::Client.new ENV["PROJECT_ID"], ENV["AUTH_TOKEN"], signalwire_space_url: ENV["SPACE_URL"]

start_date = DateTime.now - 7

# Filter by optional parameters specified in docs
calls = @client.calls.list(status: "failed", start_time_after: start_date.to_time)

# Create headers
headers = %w[SID Date Direction From To Price Status]

calls.each do |record|
  # Create and open a CSV
  CSV.open("Calls.csv", "w+") do |csv|
    # Insert headers
    csv << headers
    # For each call record, insert a row.Make sure the order of parameters matches the order of the headers, or the data will be mismatched.
    calls.each do |record|
      csv << [record.sid, record.start_time, record.direction, record.from, record.to, record.price, record.status]
    end
  end
end

File: JS/Voice/exportCallsToCSV.js
import {RestClient} from "@signalwire/compatibility-api";
import fs from "fs";
import dotenv from 'dotenv';
dotenv.config({ path: '../../.env' })

const client = new RestClient(process.env.PROJECT_ID, process.env.AUTH_TOKEN, {signalwireSpaceUrl: process.env.SPACE_URL});

// create a file and open it, create array to hold rows
let writeStream = fs.createWriteStream("Calls.csv");
let csvData = [];

// set headers, make sure header order matches order of elements in row below
const header = [
    "Call Sid",
    "To",
    "From",
    "Date",
    "Status",
    "Direction",
    "Price",
];

// push headers into array
csvData.push(header);

// list calls filtered by optional parameters from docs
client.calls
    .list({
        status: "completed",
        from: process.env.PERSONAL_NUMBER,
        startTimeAfter: new Date(Date.UTC(2022, 4, 1, 0, 0, 0)), // js datetime objects start at 0
    })
    // iterate through results and push each call record into a row
    .then((calls) => {
        calls.forEach((c) => {
            const call_record = [
                c.sid,
                c.to,
                c.from,
                c.startTime,
                c.status,
                c.direction,
                c.price
            ]
            csvData.push(call_record);
        });

        writeStream.write(csvData.join("\n"), () => {
        });
        writeStream.end();
    });
